src/db.cpp

```diff
@@ -0,0 +1,280 @@
+#include "db.h"
+#include <QtCore/QMetaObject>
+#include <QtCore/QByteArray>
+#include <QtCore/QList>
+#include <QtCore/QMap>
+#include <QtCore/QString>
+#include <QtCore/QStringList>
+#include <QtCore/QVariant>
+#include <QtDebug>
+#include <QtDBus/QtDBus>
+
+Db::Db(DBTree *dbTree, bool lookupSenderId) : dbTree(dbTree), lookupSenderId(lookupSenderId)
+{
+    // try opening xenstore handle so it's available for later use
+    xs = xs_daemon_open();
+}
+
+Db::~Db()
+{
+    // destructor
+}
+
+const QString Db::getSenderId()
+{
+    if (lookupSenderId) {
+        return message().service();
+    } else {
+        return QString("");
+    }
+}
+
+int Db::getSenderDomId()
+{
+    const QString senderId = message().service();
+
+    // use only domid == 0 if lookup is disabled
+    if (!lookupSenderId) {
+        return 0;
+    }
+
+    QDBusInterface iface("org.freedesktop.DBus", "/org/freedesktop/DBus", "org.freedesktop.DBus", QDBusConnection::systemBus());
+    if (iface.isValid()) {
+        QDBusPendingReply<int> reply = iface.asyncCall("GetConnectionDOMID", senderId);
+        reply.waitForFinished();
+        if (reply.isValid()) {
+            return reply.value();
+        }
+
+        qWarning() << "failed to read sender domid for sender id:" << senderId << "error:" << reply.error().name();
+        return -1;
+    }
+
+    qWarning() << "failed to get valid dbus interface:" << iface.lastError().name();
+    return -1;
+}
+
+QString Db::getUuidFromDomId(int domid)
+{
+    xs_transaction_t th;
+    unsigned int len;
+
+    if (xs == NULL) {
+        xs = xs_daemon_open();
+        if (xs == NULL) {
+            qWarning() << "unable to open xenstore";
+            return QString();
+        }
+    }
+
+    QString vmPathQS = "/local/domain/" + QString::number(domid) + "/vm";
+    QByteArray pba = vmPathQS.toLatin1();
+    const char *cpa = pba.data();
+    qDebug() << "performing xs_read for:" << cpa;
+
+    th = xs_transaction_start(xs);
```

Comment:
Looking over the code, I would just use XBT_NULL since your just reading. There is no need for a transaction. Then the only issue with implementation is freeing the memory as Brendan said. Probably should run some sort of tests against dbd to make sure there are no other memory leaks.